File: python/p0155.py
class MinStack:
    def __init__(self):
        self.stack = []  # Main stack to store elements
        self.min = []    # Stack to keep track of minimums

    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min or val <= self.min[-1]:
            self.min.append(val)  # Update min stack

    def pop(self) -> None:
        if self.stack:
            if self.stack[-1] == self.min[-1]:
                self.min.pop()  # Remove from min stack if it's the min value
            self.stack.pop()

    def top(self) -> int:
        if self.stack:
            return self.stack[-1]  # Return top element of the main stack

    def getMin(self) -> int:
        if self.min:
            return self.min[-1]  # Return the current minimum


# Example usage:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: javascript/p0155.js
var MinStack = function () {
  this.stack = []; // Main stack to store elements
  this.min = []; // Stack to keep track of minimums
};

/**
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function (val) {
  this.stack.push(val);
  // Update min stack if necessary
  if (this.min.length === 0 || val <= this.min[this.min.length - 1]) {
    this.min.push(val);
  }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
  if (this.stack) {
    // Remove from min stack if it's the min value
    if (this.stack[this.stack.length - 1] === this.min[this.min.length - 1]) {
      this.min.pop();
    }
    this.stack.pop();
  }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
  if (this.stack) {
    return this.stack[this.stack.length - 1]; // Return top element of the main stack
  }
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
  if (this.min) {
    return this.min[this.min.length - 1]; // Return the current minimum
  }
};

/**
 * Example usage:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */
